### PR TITLE
Atmospheric alerts computer

### PR DESCRIPTION
**Pull request by chromiumboy**
## About the PR

This PR makes the alerts computer functional. This console displays actively triggered air and fire alarms on a map.

## Why / Balance

Atmos could use some more tools to track down air leaks and fires.

## Technical details

AtmosAlertsComputerSystem.cs handles the bulk of the work. Once a second it updates a list of air and fire alarms on each grid, and passes these lists to each alerts computer. This data is used to update the appearance of the computers based on the highest alert level for that grid.  If the computer is being used, it passes the relevant alarm data to the bound UI for reporting.

## Media

[The atmospheric alerts console in action](https://youtu.be/I3HIzs7BFiM)

Screenshot
alerts-computer

- [X] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## Breaking changes

None

**Changelog**

:cl:
- add: The atmospheric alert computers are now functional! Use them to locate active air and fire alarms on the station.

### CODE REVIEW

**Comment by UbaserB** [comment]
Instead of being a dot, can it fill in the room it is assigned with in a certain colour based on the status?

**Comment by chromiumboy** [comment]
> Instead of being a dot, can it fill in the room it is assigned with in a certain colour based on the status?

Possibly. Haven't looked into it though.

**Comment by CyberDuck-Git** [comment]
Is remote gas control included?

**Comment by chromiumboy** [comment]
> Is remote gas control included?

Nope. It just shows the active alarms, and air alarms also show you a summary of their atmospheric readings

**Comment by lzk228** [comment]
will atmos pipe map be included?

**Comment by chromiumboy** [comment]
> will atmos pipe map be included?

Not on this console; I plan to have the pipe net on its own console (the atmospheric monitoring console). This will also display atmos devices, such as gas pumps and thermo regulators, and the current statuses of these devices (e.g. if they are on or off, gas flow rates, set temperature)

**Comment by chromiumboy** [comment]
Converting this to draft. Decided to wait for my nav map region PRs to be finished/merged so this console can display air alarm sensor coverage

**Comment by Emisse** [comment]
Closed due to feature freeze May 10th-June 14th. Comment to have it reopen after this.

**Comment by CyberDuck-Git** [comment]
> Closed due to feature freeze May 10th-June 14th. Comment to have it reopen after this.

I think it's time to open

**Comment by JIPDawg** [comment]
> Converting this to draft. Decided to wait for my nav map region PRs to be finished/merged so this console can display air alarm sensor coverage

Since your navmaps PR got merged, are you still working on this? If not would it be ok if I yoinked it and just PR'd as is?

**Comment by chromiumboy** [comment]
Re-opening the PR :)

**Comment by deepdarkdepths** [comment]
YAY!!!!

**Comment by JIPDawg** [comment]
YAY INDEED

**Comment by ArtisticRoomba** [comment]
every atmos player is jumping for joy I can tell you that right now
this will also be huge for AI, allowing it to deal with atmospheric faults and events with little to no human intervention (once the AI interaction component gets added to more stuff, passing MVP)

**Comment by ArtisticRoomba** [comment]
> Instead of being a dot, can it fill in the room it is assigned with in a certain colour based on the status?

would be cool, but I feel that would be too complicated and you'd get a lot of misreporting
air alarms handle their respective vents and filters, sometimes an air alarm is placed in another room to monitor and control the atmosphere of another room remotely (science xenoarch chambers, atmos burn chambers)

it would be a HUGE boost of information if we could see smaller dots that represent the various scrubbers and vents linked to said air alarm, and their statuses represented by color, this would give us a good idea as to how far an atmos emergency has spread

an even BIGGER absolutely MASSIVE bonus along with this is to indicate the position of firelocks via thin boxes on the map and whether they are open, closed, or locked (green, yellow, or red). this would give atmos an even better picture of how far a spacing or tritfire has spread, and to call it out/respond accordingly

**Comment by chromiumboy** [comment]
> > Instead of being a dot, can it fill in the room it is assigned with in a certain colour based on the status?
> 
> would be cool, but I feel that would be too complicated and you'd get a lot of misreporting air alarms handle their respective vents and filters, sometimes an air alarm is placed in another room to monitor and control the atmosphere of another room remotely (science xenoarch chambers, atmos burn chambers)
> 
> it would be a HUGE boost of information if we could see smaller dots that represent the various scrubbers and vents linked to said air alarm, and their statuses represented by color, this would give us a good idea as to how far an atmos emergency has spread
> 
> an even BIGGER absolutely MASSIVE bonus along with this is to indicate the position of firelocks via thin boxes on the map and whether they are open, closed, or locked (green, yellow, or red). this would give atmos an even better picture of how far a spacing or tritfire has spread, and to call it out/respond accordingly

I actually considered both of these ideas during development. I decided against mapping the firelocks because there are just so many (even moderately sized stations have hundreds) and it would more than double the amount of server overhead because of all the entities you have to query. I also considered having the atmos vents/sensors appear on the map, just as you said, but again it adds more overhead on the server. I also liked the idea of withholding at least some information from the players. They know roughly where the problem is, and what sort of problem they face, but will need to investigate in person to determine the extent of the issue

**Comment by ArtisticRoomba** [comment]
I agree. Good to see atmos getting some love and foundation, thank you!

**Comment by chavonadelal** [comment]
> > > Instead of being a dot, can it fill in the room it is assigned with in a certain colour based on the status?
> > 
> > 
> > would be cool, but I feel that would be too complicated and you'd get a lot of misreporting air alarms handle their respective vents and filters, sometimes an air alarm is placed in another room to monitor and control the atmosphere of another room remotely (science xenoarch chambers, atmos burn chambers)
> > it would be a HUGE boost of information if we could see smaller dots that represent the various scrubbers and vents linked to said air alarm, and their statuses represented by color, this would give us a good idea as to how far an atmos emergency has spread
> > an even BIGGER absolutely MASSIVE bonus along with this is to indicate the position of firelocks via thin boxes on the map and whether they are open, closed, or locked (green, yellow, or red). this would give atmos an even better picture of how far a spacing or tritfire has spread, and to call it out/respond accordingly
> 
> I actually considered both of these ideas during development. I decided against mapping the firelocks because there are just so many (even moderately sized stations have hundreds) and it would more than double the amount of server overhead because of all the entities you have to query. I also considered having the atmos vents/sensors appear on the map, just as you said, but again it adds more overhead on the server. I also liked the idea of withholding at least some information from the players. They know roughly where the problem is, and what sort of problem they face, but will need to investigate in person to determine the extent of the issue

What do you think about adding a display of fire gates, but only activated ones? The load on the server will not be so great, and it will be much clearer which place is isolated (ie there is no oxygen there)

**Comment by chromiumboy** [comment]
> What do you think about adding a display of fire gates, but only activated ones? The load on the server will not be so great, and it will be much clearer which place is isolated (ie there is no oxygen there)

Not entirely opposed to the idea, but I'd wait and see how the console fairs as is before making further changes

File: Content.Client/Atmos/Consoles/AtmosAlarmEntryContainer.xaml.cs [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+
+        _cache = IoCManager.Resolve<IResourceCache>();
+
+        NetEntity = uid;
+        Coordinates = coordinates;
+
+        // Load fonts
+        var headerFont = new VectorFont(_cache.GetResource<FontResource>("/Fonts/NotoSans/NotoSans-Bold.ttf"), 11);
+        var normalFont = new VectorFont(_cache.GetResource<FontResource>("/Fonts/NotoSansDisplay/NotoSansDisplay-Regular.ttf"), 11);
+        var smallFont = new VectorFont(_cache.GetResource<FontResource>("/Fonts/NotoSans/NotoSans-Regular.ttf"), 10);
```
Comment (deathride58) [inline Content.Client/Atmos/Consoles/AtmosAlarmEntryContainer.xaml.cs:58]: we know this is in-line with existing code in the codebase and all, but good LORD it'd be nice to descrunklify this

**Review by deathride58 — APPROVED** [review]

**Comment by Moomoobeef** [comment]
holy shit, mate saw the mock-up I made and actually went and made the thing. :+1:

**Comment by superjj18** [comment]
Finally the bridge is 99% feature complete(damn medical records terminal)

Absolutely amazing, now when arrivals gets pumped up to 500 degrees Celsius we can actually react and warn people about it before 5 people receive 1600 burn damage

**Pull request merged**